parla/tasks.py

```diff
@@ -447,6 +447,124 @@ def get_current_devices() -> List[Device]:
     return list(task_runtime.get_devices())
 
 
+@contextmanager
+def _reserve_persistent(memsize, device):
+    resource_pool = get_scheduler_context().scheduler._available_resources
+    resource_pool.allocate_resources(device, {'memory' : memsize}, blocking = True)
+    try:
+        yield
+    finally:
+        resource_pool.deallocate_resources(device, {'memory' : memsize})
+
+
+def _get_parla_device(device):
+    if isinstance(device, Device):
+        return device
+    try:
+        import cupy
+    except ImportError:
+        pass
+    else:
+        if isinstance(device, cupy.cuda.Device):
+            from .cuda import gpu
+            index = device.id
+            return gpu(index)
+    raise ValueError("Don't know how to convert object of type {} to a parla device object.".format(type(device)))
+
+
+@contextmanager
+def reserve_persistent_memory(amount, device = None):
+    """
+    :param amount: The number of bytes reserved in the scheduler from tasks for persitent data. \
+      This exists, not as any kind of enforced limit on allocation, but rather to let the scheduler \
+      have an accurate measure of memory occupancy on the GPU beyond just memory that's used \
+      only during a task's execution. It can be specified as an integer representing the nubmer of \
+      bytes, an ndarray (cupy or numpy), or a list of ndarrays.
+    :param device: The device object where memory is to be reserved. \
+      This must be supplied if amount is an integer \
+      and may be supplied for an array. In the case of a list or other iterable it must \
+      be supplied if any element of the list is not an array. This may be a list of \
+      devices if amount is a list of array objects.
```

Comment:
I think this is too flexible and has too many special cases. Can we split this into three functions:

* `reserve_data` takes `*args` where each is a data object (e.g., array) and reserves the size of each on it's device.
* `reserve_persistent_memory` takes a device and an amount and reserve it.
* `reserve_persistent_memories` takes a collection of pairs `(device, amount)` or dict and reserves every pair as above. (I would be OK with merging this one with the previous variant.)

As above, the memory functions would eventually be generalized into a function that can reserve any resource.

I just don't want Parla to end up line numpy with 6 different ways to call every function and different restrictions on each way to call it. It's hard to document and confusing, since functions don't have simple and concise semantics.

Reply:
While I understand that criticism I don't totally agree with it. I personally find the mental load of having tons of functions with extremely similar use-cases marginally worse than having different ways to call a single function that still, conceptually, represents one kind of operation. My main criticism of the APIs upstream is when functions cease to even represent the operation you'd expect them to represent (e.g. how numpy.transpose is the designated interface for arbitrarily permuting axes). Either way, for now I'll just make a TODO note that references this discussion and move on.

Reply:
My argument is that these different functions *don't* implement the same operation. But we will debate that later and in the end I'm fading out of the project, so my design opinions are less and less relaxant.

Reply:
Yah, I'm working toward moving on as well, so this decision can be left to others.